osquery/utils/fgets_buffer.h

```diff
@@ -0,0 +1,202 @@
+/**
+ *  Copyright (c) 2019-present, osquery Foundation
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <memory>
+#include <stdint.h>
+#include <string.h>
+#include <string>
+#include <vector>
+
+#include <osquery/filesystem/fileops.h>
+
+/*
+ * This all exists to ensure that we don't get hangs when
+ * using fgets() on pipes.  See issue #4810 for details.
+ */
+
+/**
+ * @brief Abstraction of a non blocking file for use with FgetsBuffer
+ */
+struct NonblockingFile {
+  virtual bool isValid() = 0;
+
+  virtual bool isDataAvail() = 0;
+
+  virtual ssize_t nbRead(std::vector<char>& buf) = 0;
+
+  virtual void close() = 0;
+
+  virtual ~NonblockingFile() {}
+};
+
+typedef std::shared_ptr<NonblockingFile> SPNonblockingFile;
```

Comment:
nit: is it really necessary to share ownership of the file?
What about using `std::unique_ptr` and then return a reference or a pointer to it (depending if it can be null or not) in getFile()?

Reply:
not necessary.  It's just not something I've used.

Reply:
I'm not sure I understood your answer.
In general using a `shared_ptr` is refrained unless you have a specific use case where there's no way of representing the same thing with a slightly different setup and a `unique_ptr`, or where you need to share a pointer between different threads and it's difficult to specify the resource life.

In this case I don't see why the file/pipe and the fgets_buffer itself have to be shared; SyslogEventPublisher owns everything.

Reply:
sorry for not being clear... I meant that I have never used unique_ptr, only shared_ptr.  I wasn't sure that I wanted to use this bug fix to learn unique_ptr details and edge cases, since the code has already been tested with shared_ptr.

Reply:
That said, if you wanted to make the changes, I welcome it.

Reply:
A quick rundown could be: when talking about smart pointers and acquiring resources safely a `unique_ptr` is the first go to.
It is used to represent a single owner of the resource, and when you want to pass ownership you have to `std::move` it instead of copying it (which is not possible).
It behaves exactly as a shared_ptr regarding creation/destruction except that there's no counter of "users" (or copies) to keep the resource alive.
The first time it goes out of scope it deletes the resource.

So for instance the line acquiring the pipe here https://github.com/osquery/osquery/pull/5969/files#diff-be841bec7c4dcf4a169f34e8716ab4bfR94

would become `auto pPipe = std::make_unique<NonblockingFileImpl>(FLAGS_syslog_pipe_path);` and then you `std::move` the `pPipe` into FgetsBuffer, which should take it by value.
The constructor should then `std::move` it again inside the member variable, which should be a `unique_ptr` too.
This now means that FgetsBuffer owns the resource (the NonBlockingFile) and will destruct it when going out of scope.

There's one side effect to keep into account which is that, since a unique_ptr cannot be copied, any class it's put in as a member variable cannot be copied (in this case FgetsBuffer and SyslogEventPublisher, which I don't think it makes sense to anyway?), but only moved.

Reply:
Without getting too philosophical, I put unique_ptr in the same category as move operator... there are some gotchas (e.g. copy, vague compiler errors for missing move, and [unique_ptr and incomplete types or interfaces](https://stackoverflow.com/questions/34072862/why-is-error-invalid-application-of-sizeof-to-an-incomplete-type-using-uniqu#34073093)) and very few developers know how to use them correctly.  Rust-lang has explicit ownership, so it takes care of this type of discussion from the start.  In C++, shared_ptr may allow for developers to accidentally share a pointer meant for exclusive access, it strikes the right balance.